### PR TITLE
nixos/harmonia: init service

### PR DESCRIPTION
**Pull request by Mic92**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/harmonia.nix [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      signKeyPath = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = lib.mdDoc "Path to the signing key to use for signing the cache";
+      };
+
+      settings = lib.mkOption {
+        type = lib.types.submodule {
+          freeformType = format.type;
+        };
```
Comment (Kranzes) [inline nixos/modules/services/networking/harmonia.nix:22]:
```suggestion
        inherit (format) type;
```

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/tools/package-management/harmonia/default.nix [hunk 2]
```diff
@@ -35,6 +36,7 @@ rustPlatform.buildRustPackage rec {
     updateScript = nix-update-script {
       extraArgs = [ "--version-regex" "harmonia-v(.*)" ];
     };
+    tests = { inherit (nixosTests) harmonia; };
   };
 
   meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/tools/package-management/harmonia/default.nix:42]: I just noticed that the meta section needs an update

Comment (Mic92) [reply pkgs/tools/package-management/harmonia/default.nix:42]: done.

File: pkgs/tools/package-management/harmonia/default.nix [hunk 3]
```diff
@@ -35,12 +36,13 @@ rustPlatform.buildRustPackage rec {
...
     description = "Nix binary cache";
-    homepage = "https://github.com/helsinki-systems/harmonia";
+    homepage = "https://github.com/nix-community/harmonia";
     license = licenses.mit;
-    maintainers = with maintainers; [ fab ];
+    maintainers = with maintainers; [ fab mic92 ];
```
Comment (SuperSandro2000) [inline pkgs/tools/package-management/harmonia/default.nix:46]:
```suggestion
    maintainers = with maintainers; [ mic92 ];
```

**Review by SuperSandro2000 — APPROVED** [review]
successfully deployed at two locations

**Review by Kranzes — CHANGES_REQUESTED** [review]
Can you add a package option with mkPackageOptionMD?

File: nixos/modules/services/networking/harmonia.nix [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        inherit (format) type;
+
+        description = lib.mdDoc "Settings to merge with the default configuration";
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.harmonia.settings = builtins.mapAttrs (_: v: lib.mkDefault v) {
+      bind = "[::]:5000";
```
Comment (Kranzes) [inline nixos/modules/services/networking/harmonia.nix:29]: Do you need to specify the default explicitly? Will they not be used without setting s config file?

Comment (Mic92) [reply nixos/modules/services/networking/harmonia.nix]: Done.

**Comment by Mic92** [comment]
> Can you add a package option with mkPackageOptionMD?

Done

File: nixos/modules/services/networking/harmonia.nix [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = lib.mdDoc "Path to the signing key to use for signing the cache";
+      };
+
+      package = lib.mkPackageOptionMD pkgs "harmonia" { };
+
+      settings = lib.mkOption {
+        inherit (format) type;
+
```
Comment (Kranzes) [inline nixos/modules/services/networking/harmonia.nix:23]:
```suggestion
```

File: nixos/modules/services/networking/harmonia.nix [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      requires = [ "nix-daemon.socket" ];
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      environment = {
+        NIX_REMOTE = "daemon";
+        LIBEV_FLAGS = "4"; # go ahead and mandate epoll(2)
+        CONFIG_FILE = lib.mkIf (configFile != null) configFile;
+        SIGN_KEY_PATH = lib.mkIf (cfg.signKeyPath != null) "%d/sign-key";
+        RUST_LOG = "info";
```
Comment (Kranzes) [inline nixos/modules/services/networking/harmonia.nix:42]: Maybe make a logging option? That could be helpful for debugging.

Comment (Kranzes) [reply nixos/modules/services/networking/harmonia.nix:42]: Make an option called logLevel, there are many other modules in nixpkgs that used RUST_LOG

Comment (Mic92) [reply nixos/modules/services/networking/harmonia.nix:42]: We can do this if someone actually needs this for harmonia. I don't see a use case for this particular project (there are not many debug statements in the code)

For now one can also do:

```
systemd.services.harmonia.environment.RUST_LOG = "debug";
```

Comment (Kranzes) [reply nixos/modules/services/networking/harmonia.nix:42]: Alright maybe set it as a mkDefault so people don't need to mkForce

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by vcunat** [comment]
Maybe mark `harmonia` as broken on 32-bit platforms, if noone wants to fix it?
https://hydra.nixos.org/build/219635452

**Comment by Mic92** [comment]
32-bit support fixed in https://github.com/NixOS/nixpkgs/pull/231617

**Pull request merged**